platforms/allwinner-d1/boards/lichee-rv/src/uart.rs

```diff
@@ -0,0 +1,181 @@
+        loop {
+            let rx = cons.read_grant().await;
+            let len = rx.len();
+            let thr_addr = unsafe { &*UART0::PTR }.thr() as *const _ as *mut ();
+
+            let rx_sli: &[u8] = &rx;
+
+            let d_cfg = DescriptorConfig {
+                source: rx_sli.as_ptr().cast(),
+                destination: thr_addr,
+                byte_counter: rx_sli.len(),
+                link: None,
+                wait_clock_cycles: 0,
+                bmode: BModeSel::Normal,
+                dest_width: DataWidth::Bit8,
+                dest_addr_mode: AddressMode::IoMode,
+                dest_block_size: BlockSize::Byte1,
+                dest_drq_type: DestDrqType::Uart0Tx,
+                src_data_width: DataWidth::Bit8,
+                src_addr_mode: AddressMode::LinearMode,
+                src_block_size: BlockSize::Byte1,
+                src_drq_type: SrcDrqType::Dram,
+            };
+            let descriptor = d_cfg.try_into().unwrap();
+            unsafe {
+                tx_channel.set_channel_modes(ChannelMode::Wait, ChannelMode::Handshake);
+                tx_channel.start_descriptor(NonNull::from(&descriptor));
+            }
+            let _ = TX_DONE.wait().await;
+            unsafe {
+                tx_channel.stop_dma();
+            }
+            rx.release(len);
+        }
+    }
+
+    async fn serial_server(handle: BidiHandle, kcons: KConsumer<Message<SimpleSerial>>) {
+        loop {
+            if let Ok(req) = kcons.dequeue_async().await {
+                let Request::GetPort = req.msg.body;
+                let resp = req.msg.reply_with(Ok(Response::PortHandle { handle }));
+                let _ = req.reply.reply_konly(resp).await;
+                break;
+            }
+        }
+
+        // And deny all further requests after the first
+        loop {
+            if let Ok(req) = kcons.dequeue_async().await {
+                let Request::GetPort = req.msg.body;
+                let resp = req
+                    .msg
+                    .reply_with(Err(SimpleSerialError::AlreadyAssignedPort));
+                let _ = req.reply.reply_konly(resp).await;
+            }
+        }
+    }
+
+    pub async fn register(
+        k: &'static Kernel,
+        cap_in: usize,
+        cap_out: usize,
+        tx_channel: Channel,
+    ) -> Result<(), ()> {
+        assert_eq!(tx_channel.channel_index(), 0);
+
+        let (kprod, kcons) = KChannel::<Message<SimpleSerial>>::new_async(k, 4)
+            .await
+            .split();
+        let (fifo_a, fifo_b) = new_bidi_channel(k.heap(), cap_in, cap_out).await;
+
+        let _server_hdl = k.spawn(D1Uart::serial_server(fifo_b, kcons)).await;
+
+        let (prod, cons) = fifo_a.split();
+        let _send_hdl = k.spawn(D1Uart::sending(cons, tx_channel)).await;
+
+        let boxed_prod = k.heap().allocate(prod).await;
+        let leaked_prod = boxed_prod.leak();
+        UART_RX.store(leaked_prod.as_ptr(), Ordering::Release);
```

Comment:
nit: should we maybe do a swap/compare-exchange so we can assert the previous value isn't null?

Reply:
In this case, the previous value SHOULD be null.